File: internal/spectest/runner.ts
/**
 * Test runner for cross-language SDK spec tests.
 *
 * This runner:
 * 1. Loads YAML test specifications from sdkspec/test/
 * 2. Executes SDK calls against the vendor endpoints
 * 3. Validates Braintrust API spans
 */

import * as fs from "fs";
import * as path from "path";
import * as yaml from "yaml";
import OpenAI from "openai";
import { initLogger, wrapOpenAI } from "braintrust";

interface TestSpec {
  name: string;
  description?: string;
  vendor: string;
  endpoint: string;
  request: Record<string, any>;
  braintrust_span?: Record<string, any>;
}

interface SpecFile {
  name: string;
  tests: TestSpec[];
}

/**
 * Runner for SDK specification tests.
 */
export class SpecTestRunner {
  private specPath: string;
  private spec: SpecFile;

  constructor(specPath: string) {
    this.specPath = specPath;
    this.spec = this.loadSpec();
  }

  /**
   * Load the YAML test specification.
   */
  private loadSpec(): SpecFile {
    const content = fs.readFileSync(this.specPath, "utf-8");
    return yaml.parse(content);
  }

  /**
   * Get project UUID from project name.
   */
  private async getProjectId(projectName: string): Promise<string> {
    const apiKey = process.env.BRAINTRUST_API_KEY;
    if (!apiKey) {
      throw new Error("BRAINTRUST_API_KEY environment variable not set");
    }

    const apiUrl =
      process.env.BRAINTRUST_API_URL || "https://api.braintrust.dev";

    // Fetch projects
    const url = `${apiUrl}/v1/project?project_name=${encodeURIComponent(projectName)}`;
    const headers = {
      Authorization: `Bearer ${apiKey}`,
      "Content-Type": "application/json",
    };

    const response = await fetch(url, { headers });
    if (!response.ok) {
      throw new Error(`Failed to fetch projects: ${response.statusText}`);
    }

    const data = await response.json();
    const projects = data.objects || [];

    const matching = projects.filter((p: any) => p.name === projectName);
    if (matching.length === 0) {
      throw new Error(`Project not found: ${projectName}`);
    }

    return matching[0].id;
  }

  /**
   * Fetch span with exponential backoff retry logic.
   *
   * Retries on error with backoff: 30s, then 30s increments
   * Stops when total wait time exceeds 150 seconds.
   */
  private async fetchBraintrustSpan(
    rootSpanId: string,
    projectId: string,
  ): Promise<Record<string, any>> {
    const backoffSeconds = 30;
    let totalWait = 0;
    const maxTotalWait = 150;
    let lastError: Error | null = null;

    while (true) {
      try {
        return await this.fetchBraintrustSpanImpl(rootSpanId, projectId);
      } catch (e) {
        lastError = e as Error;
        if (totalWait > maxTotalWait) {
          break;
        }
        console.log(
          `Span not found yet, waiting ${backoffSeconds}s before retry (total wait: ${totalWait}s)...`,
        );
        await new Promise((resolve) =>
          setTimeout(resolve, backoffSeconds * 1000),
        );
        totalWait += backoffSeconds;
      }
    }

    // Exceeded max wait time, re-throw the last error
    throw lastError;
  }

  /**
   * Fetch span data from Braintrust API by root_span_id using BTQL.
   *
   * Returns the child span (not the root span itself).
   */
  private async fetchBraintrustSpanImpl(
    rootSpanId: string,
    projectId: string,
  ): Promise<Record<string, any>> {
    const apiKey = process.env.BRAINTRUST_API_KEY;
    if (!apiKey) {
      throw new Error("BRAINTRUST_API_KEY environment variable not set");
    }

    const apiUrl =
      process.env.BRAINTRUST_API_URL || "https://api.braintrust.dev";

    // Use BTQL to query for all spans with this root_span_id
    const url = `${apiUrl}/btql`;
    const headers = {
      Authorization: `Bearer ${apiKey}`,
      "Content-Type": "application/json",
    };

    // Query for child spans with this root_span_id using BTQL string syntax
    // Filter for spans where root_span_id matches AND span_parents is not null (i.e., not the root)
    const btqlQuery = {
      query: `select: *\nfrom: project_logs('${projectId}')\nfilter: root_span_id = '${rootSpanId}' and span_parents != null\nlimit: 1000`,
      use_columnstore: true,
      use_brainstore: true,
      brainstore_realtime: true,
      api_version: 1,
      fmt: "json",
    };

    const response = await fetch(url, {
      method: "POST",
      headers,
      body: JSON.stringify(btqlQuery),
    });

    if (!response.ok) {
      throw new Error(`Failed to fetch span: ${response.statusText}`);
    }

    const data = await response.json();
    const childSpans = data.data || [];

    // We expect exactly 1 child span (the LLM span)
    if (childSpans.length === 0) {
      throw new Error(`No child spans found with root_span_id: ${rootSpanId}`);
    }
    if (childSpans.length !== 1) {
      throw new Error(
        `Expected exactly 1 child span, found ${childSpans.length}`,
      );
    }

    return childSpans[0];
  }

  /**
   * Validate that Braintrust span data matches expected structure.
   *
   * Recursively walks through the expected structure and validates each value
   * exists in the actual span data.
   */
  private validateBraintrustSpan(
    spanData: Record<string, any>,
    expected: Record<string, any>,
  ): void {
    if (!expected) {
      return;
    }

    const validateValue = (
      actual: any,
      expectedVal: any,
      path: string,
    ): void => {
      if (
        expectedVal !== null &&
        typeof expectedVal === "object" &&
        !Array.isArray(expectedVal)
      ) {
        // For objects, recursively validate each key
        if (typeof actual !== "object" || actual === null) {
          throw new Error(
            `Path ${path}: expected object, got ${typeof actual}`,
          );
        }
        for (const [key, val] of Object.entries(expectedVal)) {
          if (!(key in actual)) {
            throw new Error(
              `Path ${path}.${key}: key not found in actual data`,
            );
          }
          validateValue(actual[key], val, `${path}.${key}`);
        }
      } else if (Array.isArray(expectedVal)) {
        if (Array.isArray(actual)) {
          // Both are arrays - validate each element
          if (expectedVal.length !== actual.length) {
            throw new Error(
              `Path ${path}: list length mismatch, expected=${expectedVal.length}, actual=${actual.length}`,
            );
          }
          for (let i = 0; i < expectedVal.length; i++) {
            validateValue(actual[i], expectedVal[i], `${path}[${i}]`);
          }
        } else if (typeof actual === "object" && actual !== null) {
          // Expected is list of dicts (YAML format for dict key-values like metadata)
          // Actual is a dict - validate each item in the list as a key-value pair
          for (const item of expectedVal) {
            if (typeof item === "object" && item !== null) {
              for (const [key, val] of Object.entries(item)) {
                if (!(key in actual)) {
                  throw new Error(
                    `Path ${path}.${key}: key not found in actual data`,
                  );
                }
                validateValue(actual[key], val, `${path}.${key}`);
              }
            }
          }
        } else {
          throw new Error(
            `Path ${path}: expected list but actual is ${typeof actual}`,
          );
        }
      } else {
        // For scalar values, check equality or regex match
        if (
          typeof expectedVal === "string" &&
          expectedVal.startsWith("regex:")
        ) {
          // Treat as regex pattern - convert actual to string for matching
          const pattern = expectedVal.slice(6); // Remove "regex:" prefix
          const actualStr = String(actual);
          const regex = new RegExp(`^${pattern}$`);
          if (!regex.test(actualStr)) {
            throw new Error(
              `Path ${path}: regex pattern '${pattern}' did not match actual='${actualStr}' (type=${typeof actual})`,
            );
          }
        } else {
          // Exact equality check
          if (actual !== expectedVal) {
            throw new Error(
              `Path ${path}: expected=${expectedVal}, actual=${actual}`,
            );
          }
        }
      }
    };

    // Walk through each top-level key in expected
    for (const [key, expectedVal] of Object.entries(expected)) {
      if (!(key in spanData)) {
        throw new Error(`Top-level key '${key}' not found in span data`);
      }
      validateValue(spanData[key], expectedVal, key);
    }
  }

  /**
   * Run a single test from the specification.
   */
  async runTest(
    testSpec: TestSpec,
    testSuiteName: string = "Unknown",
  ): Promise<void> {
    const { vendor, endpoint, request } = testSpec;

    // Initialize Braintrust logger
    const logger = initLogger({
      projectName: "sdk-spec-test",
      setCurrent: true,
    });

    // Execute the SDK call
    let rootSpanId: string | undefined;

    await logger.traced(
      async (span) => {
        if (vendor === "OpenAI" && endpoint === "completions") {
          // Create a parent span so we can find the llm span
          const client = wrapOpenAI(new OpenAI());
          const response = await client.chat.completions.create(request as any);
          // Store root span ID for later validation
          rootSpanId = span.rootSpanId;
        } else {
          // TODO: Implement other vendor/endpoint combinations
          throw new Error(`Unsupported vendor/endpoint: ${vendor}/${endpoint}`);
        }
      },
      {
        name: `${testSuiteName}.${testSpec.name}`,
      },
    );

    // Flush to send to Braintrust API
    await logger.flush();

    // Validate Braintrust spans (if specified)
    if (testSpec.braintrust_span && rootSpanId) {
      // Fetch actual span data from Braintrust API
      // We need to get the project UUID from the project name
      // TODO: cache project id
      const projectId = await this.getProjectId("sdk-spec-test");

      // Give the API a moment to process the data
      console.log(`Waiting 30s for backend to process span ${rootSpanId}...`);
      await new Promise((resolve) => setTimeout(resolve, 30000));

      const spanData = await this.fetchBraintrustSpan(rootSpanId, projectId);
      console.log(`Got span: ${JSON.stringify(spanData, null, 2)}`);
      this.validateBraintrustSpan(spanData, testSpec.braintrust_span);
    }
  }

  /**
   * Run all tests in the specification.
   */
  async runAllTests(): Promise<void> {
    const testSuiteName = this.spec.name || "Unknown";
    const tests = this.spec.tests || [];

    console.log(`Running test suite: ${testSuiteName}`);
    console.log(`Found ${tests.length} test(s)`);

    console.log(`---- ${testSuiteName} ----`);
    for (const test of tests) {
      const testName = test.name || "Unnamed test";
      console.log(`  Running: ${testName}`);
      try {
        await this.runTest(test, testSuiteName);
        console.log(`    ✓ ${testName} passed`);
      } catch (e) {
        const error = e as Error;
        console.log(`    ✗ ${testName} failed: ${error.message}`);
        throw error;
      }
    }
  }
}

/**
 * Main entry point when run directly.
 */
async function main() {
  const args = process.argv.slice(2);

  if (args.length < 1) {
    console.log("Usage: ts-node runner.ts <path-to-spec.yaml>");
    process.exit(1);
  }

  const specPath = path.resolve(args[0]);
  const runner = new SpecTestRunner(specPath);
  await runner.runAllTests();
}

// Run if executed directly (ES module check)
if (import.meta.url === `file://${process.argv[1]}`) {
  main().catch((error) => {
    console.error("Error running tests:", error);
    process.exit(1);
  });
}

File: internal/spectest/runner.py
"""
Test runner for cross-language SDK spec tests.

This runner:
1. Loads YAML test specifications from sdkspec/test/
2. Mocks HTTP responses using the wiremock configuration
3. Executes SDK calls against the mocked endpoints
4. Captures and validates OpenTelemetry spans
5. Optionally validates Braintrust API spans
"""

import json
import os
import time
from pathlib import Path
from typing import Any, Dict, List
from unittest.mock import Mock

import braintrust
import openai
import pytest
import yaml
from braintrust import wrap_openai
from braintrust.otel import BraintrustSpanProcessor
from opentelemetry import trace
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import SimpleSpanProcessor
from opentelemetry.sdk.trace.export.in_memory_span_exporter import InMemorySpanExporter


class SpecTestRunner:
    """Runner for SDK specification tests."""

    def __init__(self, spec_path: Path):
        """Initialize the runner with a path to a YAML spec file."""
        self.spec_path = spec_path
        self.spec = self._load_spec()

    def _load_spec(self) -> Dict[str, Any]:
        """Load the YAML test specification."""
        with open(self.spec_path) as f:
            return yaml.safe_load(f)

    def _setup_otel_capture(self) -> tuple[InMemorySpanExporter, TracerProvider]:
        """Set up OpenTelemetry to capture spans in memory."""
        # Enable OTel compatibility mode for Braintrust
        os.environ['BRAINTRUST_OTEL_COMPAT'] = 'true'

        # TODO -- if python sdk ever supports otel instrumentation stop using init_logger
        braintrust.init_logger(project="sdk-spec-test")
        # Create TracerProvider and set it as global
        provider = TracerProvider()
        exporter = InMemorySpanExporter()
        provider.add_span_processor(SimpleSpanProcessor(exporter))
        provider.add_span_processor(BraintrustSpanProcessor(parent="project_name:sdk-spec-test"))
        trace.set_tracer_provider(provider)

        return exporter, provider

    def _mock_http_response(self, wiremock_config: Dict[str, Any]) -> Mock:
        """Create a mock HTTP response from wiremock configuration."""
        response_config = wiremock_config["response"]
        mock_response = Mock()
        mock_response.status_code = response_config["status"]
        mock_response.headers = response_config.get("headers", {})
        mock_response.json.return_value = response_config["body"]
        mock_response.text = json.dumps(response_config["body"])
        return mock_response

    def _validate_otel_span(self, spans: List[Any], expected: Dict[str, Any]) -> None:
        """Validate that captured OTel spans match expected attributes."""
        if not expected:
            return

        span_name = expected.get("span_name")
        required_attrs = expected.get("required_attributes", [])

        # Find span by name if specified
        matching_spans = spans
        if span_name:
            matching_spans = [s for s in spans if s.name == span_name]
            if len(matching_spans) == 0:
                # Format captured span data for error message
                captured_span_info = []
                for span in spans:
                    span_dict = {
                        "name": span.name,
                        "attributes": dict(span.attributes) if hasattr(span, "attributes") else {},
                        "status": str(span.status) if hasattr(span, "status") else None,
                    }
                    captured_span_info.append(span_dict)

                error_msg = f"No span found with name: {span_name}\n\n"
                error_msg += f"Captured {len(spans)} span(s):\n"
                error_msg += json.dumps(captured_span_info, indent=2)
                assert False, error_msg

        # Validate required attributes
        for attr_config in required_attrs:
            attr_list = attr_config.get("attribute", [])
            # Parse attribute name and value
            attr_name = next((item["name"] for item in attr_list if "name" in item), None)
            attr_value = next((item["value"] for item in attr_list if "value" in item), None)

            if attr_name and attr_value:
                found = False
                for span in matching_spans:
                    if hasattr(span, "attributes") and span.attributes.get(attr_name) == attr_value:
                        found = True
                        break
                assert found, f"Required attribute not found: {attr_name}={attr_value}"

    def _get_project_id(self, project_name: str) -> str:
        """Get project UUID from project name."""
        import requests

        api_key = os.getenv("BRAINTRUST_API_KEY")
        if not api_key:
            raise ValueError("BRAINTRUST_API_KEY environment variable not set")

        api_url = os.getenv("BRAINTRUST_API_URL", "https://api.braintrust.dev")

        # Fetch projects
        url = f"{api_url}/v1/project"
        headers = {
            "Authorization": f"Bearer {api_key}",
            "Content-Type": "application/json"
        }

        response = requests.get(url, headers=headers, params={"project_name": project_name})
        response.raise_for_status()

        data = response.json()
        projects = data.get("objects", [])

        matching = [p for p in projects if p.get("name") == project_name]
        if not matching:
            raise ValueError(f"Project not found: {project_name}")

        return matching[0]["id"]

    def _fetch_braintrust_span(self, root_span_id: str, project_id: str) -> Dict[str, Any]:
        """Fetch span data from Braintrust API by root_span_id.

        Returns the child span (not the root span itself).
        """
        import requests

        api_key = os.getenv("BRAINTRUST_API_KEY")
        if not api_key:
            raise ValueError("BRAINTRUST_API_KEY environment variable not set")

        api_url = os.getenv("BRAINTRUST_API_URL", "https://api.braintrust.dev")

        # Fetch all events with this root_span_id
        url = f"{api_url}/v1/project_logs/{project_id}/fetch"
        headers = {
            "Authorization": f"Bearer {api_key}",
            "Content-Type": "application/json"
        }

        # Fetch with limit - we expect root span + 1 child
        response = requests.post(url, headers=headers, json={"limit": 10})
        response.raise_for_status()

        data = response.json()
        events = data.get("events", [])

        # Filter events by root_span_id
        trace_events = [e for e in events if e.get("id") == root_span_id]

        if len(trace_events) == 0:
            raise ValueError(f"No events found with root_span_id: {root_span_id}")

        # Find the child span (not the root)
        child_spans = [e for e in trace_events if not e.get("is_root", False)]

        if len(child_spans) == 0:
            raise ValueError(f"No child spans found for root_span_id: {root_span_id}")

        if len(child_spans) > 1:
            print(f"Warning: Found {len(child_spans)} child spans, expected 1. Using first.")

        return child_spans[0]

    def _validate_braintrust_span(self, span_data: Dict[str, Any], expected: Dict[str, Any]) -> None:
        """Validate that Braintrust span data matches expected structure."""
        if not expected:
            return

        # Validate metadata
        expected_metadata = expected.get("metadata", [])
        for meta_item in expected_metadata:
            for key, value in meta_item.items():
                assert span_data.get("metadata", {}).get(key) == value, \
                    f"Metadata mismatch: {key}={value}"

        # Validate span attributes
        expected_attrs = expected.get("span_attributes", {})
        for key, value in expected_attrs.items():
            assert span_data.get(key) == value, f"Span attribute mismatch: {key}={value}"

    def run_test(self, test_spec: Dict[str, Any]) -> None:
        """Run a single test from the specification."""
        vendor = test_spec["vendor"]
        endpoint = test_spec["endpoint"]
        request = test_spec["request"]
        wiremock_str = test_spec.get("wiremock", "{}")
        wiremock_config = json.loads(wiremock_str)

        # Set up OTel capture
        exporter, provider = self._setup_otel_capture()

        # Mock the HTTP response
        mock_response = self._mock_http_response(wiremock_config)

        root_span_id = ""
        # Execute the SDK call
        if vendor == "OpenAI" and endpoint == "completions":
            # Initialize Braintrust logger
            logger = braintrust.init_logger(project="sdk-spec-test", set_current=True)

            # Create a parent span to capture the trace
            with logger.start_span(name=test_spec["name"]) as parent_span:
                # Make the API call (will be automatically traced as child span)
                client = wrap_openai(openai.OpenAI())
                response = client.chat.completions.create(**request)

            # Flush to send to Braintrust API
            logger.flush()

            # Store the span ID for fetching trace data
            root_span_id = parent_span.id
            # print(f"Trace ID: {trace_id}")
            # print(f"Permalink: {parent_span.permalink()}")
        else:
            # TODO: Implement other vendor/endpoint combinations
            pass

        # Validate OTel spans
        captured_spans = exporter.get_finished_spans()
        # NOTE: python sdk does not instrument in otel so we'll skip otel validation
        # if "otel_span" in test_spec:
        #     self._validate_otel_span(captured_spans, test_spec["otel_span"])

        # Validate Braintrust spans (if specified)
        if "braintrust_span" in test_spec and root_span_id:
            # Fetch actual span data from Braintrust API
            # We need to get the project UUID from the project name
            # TODO cache project id
            project_name = "sdk-spec-test"
            project_id = self._get_project_id(project_name)

            # Give the API a moment to process the data
            time.sleep(5)
            span_data = self._fetch_braintrust_span(root_span_id, project_id)
            self._validate_braintrust_span(span_data, test_spec["braintrust_span"])

    def run_all_tests(self) -> None:
        """Run all tests in the specification."""
        test_name = self.spec.get("name", "Unknown")
        tests = self.spec.get("tests", [])

        print(f"Running test suite: {test_name}")
        print(f"Found {len(tests)} test(s)")

        for test in tests:
            test_name = test.get("name", "Unnamed test")
            print(f"  Running: {test_name}")
            try:
                self.run_test(test)
                print(f"    ✓ {test_name} passed")
            except AssertionError as e:
                print(f"    ✗ {test_name} failed: {e}")
                raise


def pytest_collect_file(parent, file_path):
    """Pytest hook to collect YAML spec files as tests."""
    if file_path.suffix == ".yaml" and "sdkspec/test" in str(file_path):
        return SpecTestFile.from_parent(parent, path=file_path)


class SpecTestFile(pytest.File):
    """Pytest file node for YAML spec files."""

    def collect(self):
        """Collect individual tests from the YAML spec."""
        runner = SpecTestRunner(self.path)
        for test_spec in runner.spec.get("tests", []):
            yield SpecTestItem.from_parent(self, name=test_spec["name"], runner=runner, test_spec=test_spec)


class SpecTestItem(pytest.Item):
    """Pytest item representing a single spec test."""

    def __init__(self, name, parent, runner, test_spec):
        super().__init__(name, parent)
        self.runner = runner
        self.test_spec = test_spec

    def runtest(self):
        """Execute the test."""
        self.runner.run_test(self.test_spec)

    def repr_failure(self, excinfo):
        """Represent test failures."""
        return f"Spec test failed: {excinfo.value}"

    def reportinfo(self):
        """Report test information."""
        return self.fspath, 0, f"spec: {self.name}"


if __name__ == "__main__":
    # Can be run directly for debugging
    import sys

    if len(sys.argv) < 2:
        print("Usage: python runner.py <path-to-spec.yaml>")
        sys.exit(1)

    spec_path = Path(sys.argv[1])
    runner = SpecTestRunner(spec_path)
    runner.run_all_tests()
